### PR TITLE
Initial proposal for Device Swap API

### LINKED ISSUE #6: Discuss API Design
**Problem description**
Potentially Device Swap API could provided 2 distinct information:

-  Request last Device Swap date (_When was the last device swap for the mobile line identified by the phone number +123456799?_)
- Check last Device swap by providing a delay (_Does we had a device swap for the last 240 hours for the mobile line identified by the phone number +123456799?_)

In order to move forward on the API design we should discuss on following questions

1. Are we sharing same expectations of the API?
2. Are we considering this information as independent one from the other ? We can have different privacy requirement, different way to commercialize it. Technical implementation feasibility should be considered.
3. How multi-SIM can affect this design?  

**Possible evolution**

**Alternative solution**

**Additional context**

### PR DESCRIPTION
**Pull request by jgarciahospital**
#### What type of PR is this?

Add one of the following kinds:

* enhancement/feature
* documentation

#### What this PR does / why we need it:

First API proposal, including both endpoint functionality as part of the same API yaml.

Documentation will need to also include detailed description for the kind of scenarios where a Device Swap is considered, per #11 discussion.

#### Which issue(s) this PR fixes:

Fixes #6

### CODE REVIEW

File: code/API_definitions/device-swap.yaml [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+                $ref: "#/components/schemas/CheckDeviceSwapInfo"
+        "400":
+          $ref: "#/components/responses/Generic400Check"
+        "401":
+          $ref: "#/components/responses/Generic401"
+        "403":
+          $ref: "#/components/responses/Generic403"
+        "404":
+          $ref: "#/components/responses/Generic404"
+        "429":
```
Comment (bigludo7) [inline code/API_definitions/device-swap.yaml:163]: For consistency with other API and in particular with SIM swap I suggest to add error 422

```
    Generic422:
      description: Unprocessable Content
      headers:
        x-correlator:
          $ref: "#/components/headers/x-correlator"
      content:
        application/json:
          schema:
            $ref: "#/components/schemas/ErrorInfo"
          examples:
            GENERIC_422_NOT_SUPPORTED:
              description: Not Supported
              value:
                status: 422
                code: NOT_SUPPORTED
                message: Service not supported for this phoneNumber
            UNIDENTIFIABLE_PHONE_NUMBER:
              description: The phone number is not included in the request and the phone number information cannot be derived from the 3-legged access token
              value:
                status: 422
                code: UNIDENTIFIABLE_PHONE_NUMBER
                message: The phone number cannot be identified`

File: code/API_definitions/device-swap.yaml [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+          $ref: "#/components/responses/Generic429"
```
Comment (bigludo7) [inline code/API_definitions/device-swap.yaml:164]: 429 is not defined for sims swap.
I'm fine to have it here but should we add it in simswap? cc @fernandopradocabrillo

File: code/API_definitions/device-swap.yaml [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+                $ref: "#/components/schemas/DeviceSwapInfo"
+        "400":
+          $ref: "#/components/responses/Generic400"
+        "401":
+          $ref: "#/components/responses/Generic401"
+        "403":
+          $ref: "#/components/responses/Generic403"
+        "404":
+          $ref: "#/components/responses/Generic404"
+        "429":
```
Comment (bigludo7) [inline code/API_definitions/device-swap.yaml:116]: See comment for 422 & 429 below as same apply here.

File: code/API_definitions/device-swap.yaml [hunk 4]
```diff
@@ -0,0 +1,399 @@
+openapi: 3.0.3
+info:
+  title: Device Swap
+  description: |
+      This API allows to check the last time that the phone (device), where a mobile number (MSISDN) is installed, has changed.
```
Comment (bigludo7) [inline code/API_definitions/device-swap.yaml:5]: not sure about "a mobile number (MSISDN) is installed" meaning. What about something simpler as "This API allows to check the last time that the phone (device) - phone number association has changed"

**Review by bigludo7 — COMMENTED** [review]
Thanks
Look globally good for me. Just a small request to add 422 for alignement with other API.

File: code/API_definitions/device-swap.yaml (hunk 1 above)
Comment (jgarciahospital) [reply code/API_definitions/device-swap.yaml]: included

File: code/API_definitions/device-swap.yaml (hunk 3 above)
Comment (jgarciahospital) [reply code/API_definitions/device-swap.yaml]: both solved

File: code/API_definitions/device-swap.yaml (hunk 4 above)
Comment (jgarciahospital) [reply code/API_definitions/device-swap.yaml]: modified description

**Review by bigludo7 — COMMENTED** [review]
Strange - I've a parser error on my favorite swagger ui for lines 62-63 :(

Should be:
```
  license:
    name: Apache 2.0
    url: https://www.apache.org/licenses/LICENSE-2.0.html
  version: wip
  x-camara-commonalities: 0.4.0

**Comment by jgarciahospital** [comment]
> Strange - I've a parser error on my favorite swagger ui for lines 62-63 :(
> 
> Should be:
> 
> ```
>   license:
>     name: Apache 2.0
>     url: https://www.apache.org/licenses/LICENSE-2.0.html
>   version: wip
>   x-camara-commonalities: 0.4.0
> ```

Solved... indentation typo

**Review by bigludo7 — APPROVED** [review]
LGTM

**Comment by jgarciahospital** [comment]
Thanks ludovic for the review!

Should we merge and continue the discussion, in another PR, about the API description (as in https://github.com/camaraproject/DeviceSwap/discussions/11)?

**Comment by bigludo7** [comment]
Let keep this week for @krishvenkatachalam to provide feedback?
If not feedback we merge it Friday  by EOB

Does it work for you Jorge?

**Pull request merged**